From 400a9b3222bf86948e5a98020375c28f42eb7b51 Mon Sep 17 00:00:00 2001
From: Banka Sai Krishna <quic_banksaik@quicinc.com>
Date: Wed, 21 Aug 2024 15:52:11 +0530
Subject: [PATCH] QSDK: hostapd: Fix center frequency after channel switch

Currently, when AP operating on 5 GHz band and
performs a switch to 240 MHz bandwidth causes
stations to associate in 40 MHz instead of 160 MHz.

This is because center frequency is not updated properly
after channel switch and filling invalid CCSF values in
VHT operation IE caused the station s to use 40 MHz
bandwidth.

Fix this issue by updating center frequency properly,
during a channel / bandwidth switch.

Signed-off-by: Banka Sai Krishna <quic_banksaik@quicinc.com>
---
 src/ap/dfs.c | 3 +++
 1 file changed, 3 insertions(+)

--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -1403,6 +1403,8 @@ int hostapd_dfs_complete_cac(struct host
 	}
 
 	if (success) {
+		u8 seg0;
+
 		/* Complete iface/ap configuration */
 		if (iface->drv_flags & WPA_DRIVER_FLAGS_DFS_OFFLOAD) {
 			/* Complete AP configuration for the first bring up. */
@@ -1449,6 +1451,8 @@ int hostapd_dfs_complete_cac(struct host
 			 */
 			if (iface->state != HAPD_IFACE_ENABLED &&
 			    hostapd_is_dfs_chan_available(iface)) {
+				ieee80211_freq_to_chan(cf1, &seg0);
+				hostapd_set_oper_centr_freq_seg0_idx(iface->conf, seg0);
 				hostapd_setup_interface_complete(iface, 0);
 				iface->cac_started = 0;
 			}
